README.md

```diff
@@ -10,46 +10,180 @@ Stage: 1
 
 ## Proposal
 
-The proposal is to permit an optional string literal attribute to be associated
-with any import:
+This proposal enables reflection capabilities for both JavaScript and
+WebAssembly modules by enabling a new type of import, an import reflection:
 
 ```js
 import x from "<specifier>" as "<reflection-type>";
 ```
 
-Here the import reflection type is added to the end of the ImportStatement,
+The type of the reflection object for WebAssembly would be a
+`WebAssembly.Module`, as defined in the
+[WebAssembly JS integration API][wasm-js-api].
+
+The reflection would represent an unlinked and uninstantiated module, while
+still being able to support the same CSP policy as the native
+[ESM integration][wasm-esm], avoiding the need for `unsafe-wasm-eval` for
+custom Wasm execution.
+
+The type of the reflection object for a JavaScript module would be defined
+explicitly in this specification as a new global, `SourceTextModule`, a class
+with the following interface:
+
+```js
+class SourceTextModule {
+  // create a new instance of this module, with the given global environment record
+  instantiate (globalEnvironmentRecord: Record<string, Binding>): ModuleInstance;
+  
+  // static function to retrieved the named exports of the module.
+  // name is the name of the export, and '*' in the case of a star reexport,
+  // along with the module specifier being reexported
+  static exports(): ({ name: string } | { module: string, name: '*' })[];
+
+  // static function to retrieve imported binding list, where "module"
+  // corresponds to the module specifier being imported.
+  // star reexports appear as both exports and imports
+  // reexports also appear as both export entries for the named reexports
+  // as well as imports
+  static imports(): { module: string }[];
+
+  // possible reference to uninstantiated ModuleInstance class as a global
+  // handle for instanceof checks and other reflection use cases.
+  static class ModuleInstance;
+}
+```
+
+`SourceTextModule.exports(module)` and `SourceTextModule.imports(module)` can
+be used to inspect the module imports and exports, analogously to
+`WebAssembly.Module.exports` and `WebAssembly.Module.imports`, and in fact
+conform to the same API conventions such that code analyzing Wasm and JS
+can use similar checks.
+
+The `instantiate` method of the `SourceTextModule` class takes a global
+environment record and always returns an _unlinked_ `ModuleInstance` class
+instance with the following interface:
+
+```js
+class ModuleInstance {
+  // per current spec record state
+  state: 'unlinked' | 'linked' | 'evaluating' | 'evaluating-async' | 'evaluated' | 'errored';
+  // link the dependency specifiers to their instances
+  link (importObj: {
+    [specifier: string]: ModuleInstance | WebAssembly.Instance
+  });
+  // evaluate an instance
+  evaluate (): void | Promise<void>;
+  // access the namespace once evaluated
+  exports: null | ModuleNamespaceExoticObject;
```

Comment:
this seems kind of weird; why isn't `evaluate` returning a Promise for the exports?

Reply:
Either approach can work certainly, although it could be useful to keep the instance around as the primary source of truth over calling evaluate and throwing away the ModuleInstance instance.

This is symmetrical with `WebAssembly.Instance.prototype.exports` as well - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Instance.

Reply:
Alternatively we could call it `.namespace` and then specify a `WebAssembly.Instance.prototype.namespace` which exactly matches the exotic object for better unification - these kinds of details are more later phase decisions though and I don't think affect the overall model right now.

Reply:
I think a data property that changes with spooky action-at-a-distance is a very poorly designed API. I think it's far better if the only way to get access to an async thing is inside the Promise that represents the asynchronous work.

Reply:
It's a stateful getter as opposed to a data property though. The module instance is a stateful object that transitions through various states, including a possible error state (as exposed via the `state` getter). For example, we may also want to expose an `error` getter as well for the error state.

It can also be useful to get namespaces for other modules in the graph this way (that weren't the original target of `evaluate`).

Reply:
oof, stateful objects sound like something to avoid :-/

Reply:
@ljharb can you elaborate on the concerns here? Module records in ECMA262 are already stateful since modules go through state transitions as part of the evaluation cycle. We could hide the state, but that would be choosing deliberately removing information from the reflection system - a decision that should be made based on good reasons not to expose the state as well.

Reply:
Exposing the module record is fine; I’m saying i don’t want additional mutable state on top of that.

Reply:
`[[Namespace]]`, `[[Status]]`, `[[EvaluationError]]` are already internal fields on module records, where `[[Namespace]]` is initially empty until the module is linked. We could make the namespace available after linking but before evaluation to exactly match the current module records - the point being this is only exposing existing state, not creating new mutable state.

Reply:
I've updated the proposal to use the term `namespace` and make it available from instantiation time so it is always available as the `ModuleNamespace` and its availability is not dependent on any timing or other mutable state.

All fields are not just getters to representations of the internal module state.

Reply:
Edit: *now just getters